src/Lean/Message.lean

```diff
@@ -505,6 +528,56 @@ def indentD (msg : MessageData) : MessageData :=
 def indentExpr (e : Expr) : MessageData :=
   indentD e
 
+/--
+Returns the character length of the message when rendered.
+
+Note: this is a potentially expensive operation that is only relevant to message data that are
+actually rendered. Consider using this function in lazy message data to avoid unnecessary
+computation for messages that are not displayed.
+-/
+private def MessageData.formatLength (ctx : PPContext) (msg : MessageData) : BaseIO Nat := do
+  let { env, mctx, lctx, opts, ..} := ctx
+  let fmt ← msg.format (some { env, mctx, lctx, opts })
+  return fmt.pretty.length
+
+
+/--
+Renders an expression `e` inline in a message unless it will exceed `maxInlineLength` characters, in
+which case the expression is indented on a new line.
+
+Note that the output of this function is formatted with preceding and trailing space included. Thus,
+in `m₁ ++ inlineExpr e ++ m₂`, `m₁` should not end with a space or new line, nor should `m₂` begin
+with one.
+-/
+def inlineExpr (e : Expr) (maxInlineLength := 30) : MessageData :=
+  .ofLazy
+    (fun ctx? => do
+      let msg := MessageData.ofExpr e
+      let some ctx := ctx? | return Dynamic.mk (" " ++ msg ++ " ")
+      if (← msg.formatLength ctx) > maxInlineLength then
+        return Dynamic.mk (indentD msg ++ "\n")
+      else
+        return Dynamic.mk (" " ++ msg ++ " "))
+    (fun mctx => instantiateMVarsCore mctx e |>.1.hasSyntheticSorry)
+
+/--
+Produces a message `e : ty` that moves `: ty` to a second, indented line if a maximum length is
+exceeded.
+-/
+def indentExprWithType (e : Expr) (ty : Expr) (maxSingleLineLength := 30) : MessageData :=
+  .ofLazy
+    (fun ctx? => do
+      let eMsg := MessageData.ofExpr e
+      let tyMsg := MessageData.ofExpr ty
+      let msg := eMsg ++ " : " ++ tyMsg
+      let some ctx := ctx? | return Dynamic.mk msg
+      if (← msg.formatLength ctx) ≤ maxSingleLineLength then
+        return Dynamic.mk msg
+      else
+        return Dynamic.mk <| eMsg ++ indentD (": " ++ tyMsg))
+    (fun mctx => (instantiateMVarsCore mctx e).1.hasSyntheticSorry ||
+                 (instantiateMVarsCore mctx ty).1.hasSyntheticSorry)
```

Comment:
Based on the name I’d assume that I can use `indentExprWithType` like `indentExpr`. But unless I am misreading the code, if it fits into the line, it’s not indented? So is it more `inlineExprWithType`? What if `eMsg` itself is too long?

Reply:
This was a function I was asked to add while reviewing the error message style guidance, so I'm not totally certain about its motivations. There's nothing immediately pending that depends on it, so perhaps it's best for me to remove it from this PR and add it in a future one?

Reply:
In that case maybe defer. Also because maybe the function should just have an `Expr` parameter, and no or only optional `ty`  parameter, and infer it itself, for more convenience in the common case.

Reply:
Sounds good; I've removed this function.